crates/keyshare/src/threshold_share_collector.rs

```diff
@@ -7,75 +7,150 @@
 use std::{
     collections::{HashMap, HashSet},
     sync::Arc,
-    time::Instant,
+    time::{Duration, Instant},
 };
 
-use actix::{Actor, Addr, Handler};
-use e3_events::{ThresholdShare, ThresholdShareCreated};
+use actix::{Actor, ActorContext, Addr, AsyncContext, Handler, Message, SpawnHandle};
+use e3_events::{E3id, ThresholdShare, ThresholdShareCollectionFailed, ThresholdShareCreated};
 use e3_trbfv::PartyId;
-use tracing::info;
+use tracing::{info, warn};
 
 use crate::{AllThresholdSharesCollected, ThresholdKeyshare};
 
-pub enum CollectorState {
+const DEFAULT_COLLECTION_TIMEOUT: Duration = Duration::from_secs(120);
+
+pub(crate) enum CollectorState {
     Collecting,
     Finished,
+    TimedOut,
 }
 
+/// Message sent when threshold share collection times out.
+#[derive(Message, Clone, Debug)]
+#[rtype(result = "()")]
+pub struct ThresholdShareCollectionTimeout;
+
 pub struct ThresholdShareCollector {
+    e3_id: E3id,
     todo: HashSet<PartyId>,
     parent: Addr<ThresholdKeyshare>,
     state: CollectorState,
     shares: HashMap<PartyId, Arc<ThresholdShare>>,
+    timeout_handle: Option<SpawnHandle>,
 }
 
 impl ThresholdShareCollector {
-    pub fn setup(parent: Addr<ThresholdKeyshare>, total: u64) -> Addr<Self> {
-        let addr = Self {
+    pub fn setup(parent: Addr<ThresholdKeyshare>, total: u64, e3_id: E3id) -> Addr<Self> {
+        let collector = Self {
+            e3_id,
             todo: (0..total).collect(),
             parent,
             state: CollectorState::Collecting,
             shares: HashMap::new(),
-        }
-        .start();
-        addr
+            timeout_handle: None,
+        };
+        collector.start()
     }
 }
 
 impl Actor for ThresholdShareCollector {
     type Context = actix::Context<Self>;
+
+    fn started(&mut self, ctx: &mut Self::Context) {
+        info!(
+            e3_id = %self.e3_id,
+            "ThresholdShareCollector started, scheduling timeout in {:?}",
+            DEFAULT_COLLECTION_TIMEOUT
+        );
+        // Schedule timeout
+        let handle = ctx.notify_later(ThresholdShareCollectionTimeout, DEFAULT_COLLECTION_TIMEOUT);
+        self.timeout_handle = Some(handle);
```

Comment:
Not blocking this but I am not a huge fan of using actix for this unless there is a good reason not to use tokio::spawn as it limits the portability of our code - although we could argue that we are married to actix at this point.